src/memory.rs

```diff
@@ -1,6 +1,59 @@
 use std::mem;
 use std::slice;
 
+/// A view into an externally owned byte slice (Go `[]byte`).
+/// Use this for the current call only. A view cannot be copied for safety reasons.
+/// If you need a copy, use [`to_owned`].
+///
+/// Go's nil value is fully supported, such that we can differentiate between nil and an empty slice.
+#[repr(C)]
+pub struct ByteSliceView {
+    /// True if and only if the byte slice is nil in Go. If this is true, the other fields must be ignored.
+    is_nil: bool,
+    ptr: *const u8,
+    len: usize,
+}
+
+impl ByteSliceView {
+    /// ByteSliceViews are only constructed in Go. This constructor is a way to mimic the behaviour
+    /// when testing FFI calls from Rust. It must not be used in production code.
+    #[cfg(test)]
+    pub fn new(source: &[u8]) -> Self {
+        Self {
+            is_nil: false,
+            ptr: source.as_ptr(),
+            len: source.len(),
+        }
+    }
+
+    /// ByteSliceViews are only constructed in Go. This constructor is a way to mimic the behaviour
+    /// when testing FFI calls from Rust. It must not be used in production code.
+    #[cfg(test)]
+    pub fn nil() -> Self {
+        Self {
+            is_nil: true,
+            ptr: std::ptr::null::<u8>(),
+            len: 0,
+        }
+    }
+
+    /// Provides a reference to the included data to be parsed or copied elsewhere
+    /// This is safe as long as the `ByteSliceView` is constructed correctly.
+    pub fn read(&self) -> Option<&[u8]> {
+        if self.is_nil {
+            None
+        } else {
+            Some(unsafe { slice::from_raw_parts(self.ptr, self.len) })
```

Comment:
Question: why do you hide the `unsafe` here, rather than piping it up.
If those are bad values, this could be unsafe. Generally if you function cannot ensure the values passed are correct, you can keep the function as unsafe, so the caller knows to be careful.

Reply:
The idea is: ByteSliceView cannot be constructed in Rust code as it has no public constructor. As a consequence, all usages of `ByteSliceView::read` are safe assuming they are created correctly on the Go side. But then again, I don't know for sure how "safe" is defined.

Reply:
Everything touching FFI is implicitly unsafe. Only once it has been transfered into Rust can we handle it safely.

Reply:
This is probably true. However, bubbling up the unsafety makes this whole project unsafe. E.g.

```rust
pub fn to_cache(ptr: *mut cache_t) -> Option<&'static mut Cache<GoApi, GoStorage, GoQuerier>> {
    if ptr.is_null() {
        None
    } else {
        let c = unsafe { &mut *(ptr as *mut Cache<GoApi, GoStorage, GoQuerier>) };
        Some(c)
    }
}
```

is not a safe opertion and we have no way to tell of the memory under this pointer is really the correct type. This makes `to_cache` unsafe, which makes all the C exports unsafe.

Then if we get a

```
#[derive(Copy, Clone)]
#[repr(C)]
pub struct Buffer {
    pub ptr: *mut u8,
    pub len: usize,
    pub cap: usize,
}
```

from FFI, we cannot trust any of it's fields. So not only `Buffer::read` is unsafe, but the whole structure and everything that depends on it.

Unfortunately, I don't have a great idea how to encapsulate all this unsafety and don't know if it is even possible to make it safe.

Reply:
Fair enough. Leave it how you did it.

I think I did something similar until Reuven gave me a lecture on unsafe. It definitely doesn't change the unsafe assumptions, just adds more keywords